### PR TITLE
Add java code implementation example of template pattern

### PR DESCRIPTION
**Pull request by Jessica-Pelletier**

### CODE REVIEW

**Comment by abhiram-shaji** [comment]
Hi Jessica, this looks pretty impressive and well explained. However, I would like to point out an issue with your package name. In Java, package names are written in all lower case to avoid conflict with the names of classes or interfaces. Also, I think you named it 'observal' instead of 'template' by mistake.

**Comment by Jessica-Pelletier** [comment]
Hi @abhiram-shaji 
Thanks for looking this over for me! I see these issues now. The package name auto filled for me because of the file structure.  Do you think it makes sense to simply change the package name from patterns.Behavioral.Observal.Java -> patterns.behavior.template.java?  I may have to look into this further.

**Comment by abhiram-shaji** [comment]
> Thanks for looking this over for me! I see these issues now. The package name auto filled for me because of the file structure. Do you think it makes sense to simply change the package name from patterns.Behavioral.Observal.Java -> patterns.behavior.template.java? I may have to look into this further.

That definitely seem to be more suitable package name

File: patterns/Behavioral/Observal/Java/template.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    void processData() {
+        System.out.println("Processing Data from database: " + databaseName);
+    }
+}
+
+public class Main {
+    public static void main(String[] args) {
+        // Create an instance of FileDataProcessor
+        DataProcessor fileProcessor = new FileDataProcessor("data.txt");
+        System.out.println("Processing file data");
```
Comment (ash-teach) [inline patterns/Behavioral/Observal/Java/template.java]: These `println` statements in `main` are probably not necessary, given that there is plenty of output already.

File: patterns/Behavioral/Observal/Java/template.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        System.out.println("Save the processed data");
+    }
+
+    // The abstract method to be implemented by subclasses
+    abstract void processData();
+
+}
+
+/* The subclass 'FileDataProcessor' provide their own implementations of the 'processData()' method. Each tailered to handle data from their individual data source. When 'process()' is called on an instance of 'fileDataProcessor' it executes the steps in the template method. Utilizing the concrete implementations provided by the subclasses. */
+
```
Comment (ash-teach) [inline patterns/Behavioral/Observal/Java/template.java]: Some typos: "tailored" and "**F**ileDataProcessor"

File: patterns/Behavioral/Observal/Java/template.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
+/* The Template Method Pattern is a design pattern that defines a skeleton or outline of an algorithm in the superclass. The superclass provides a default implementation of the algorithm, but specific steps can be overridden by subclasses to customize behavior without altering the main structure. The superclass is typically not final, allowing for extension, and the pattern enables code reuse and flexibility in algorithm design. */
+
+/* 
+ * This in an example of a data processing program. It defines the skeleton of the algorithm and sets the 'process' method to final to prohibit changes being made to the outline. This base class is abstract class to allow for the methods to be implemented by subclasses. In this case the 'process' method is the template method and outlines the order of execution.
+ */
```
Comment (ash-teach) [inline patterns/Behavioral/Observal/Java/template.java]: If you have used any external resources, could you please supply them as references here as well?

**Review by ash-teach — CHANGES_REQUESTED** [review]

**Comment by ash-teach** [comment]
Great work, @Jessica-Pelletier. I've noted a couple minor fixes above, if you can address them.

**Comment by Jessica-Pelletier** [comment]
Changes have been made ! I got a little confused in the process of pushing my branch of my fork. Hopefully it's all good.

**Pull request merged**